LIBCLOUD-450: Update GCE to v1

wrigri opened the pull request:
Update GCE to v1.
The primary change for v1 from v1beta16 is that ephemeral disks have gone away and now all disks are persistent.  This means that create_node is now creating a disk first if an existing disk isn't supplied.
Also, add support for snapshots.

This is obviously a fairly large change, although most of the actual changed lines are for test fixtures.  Now that v1 has launched, the treadmill of releases with large changes should slow down to a more reasonable pace.

micolous committed:
LIBCLOUD-433: Make storage backend send content-type of
"application/octet-stream" if none is supplied and none can be guessed.

Also update affected tests.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Modify CloudStack driver to correctly throw InvalidCredsError exception if
invalid credentials are provided.

Kami committed:
Add new driver for Ikoula public cloud (http://express.ikoula.co.uk/cloudstack).

jobelenus committed:
service type has changed for cloudFilesCDN

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Add tests for libcloud.utils.networking.is.valid_ip_address.

Kami committed:
Add tests for is_public_subnet and is_private_subnet function.

Kami committed:
Update Rackspace auth 2.0 fixtures to reflect a new value for the "type"
attribute of the CND service catalog entry.

Kami committed:
Update CHANGES.

wrigri committed:
Update API to v1 and remove traces of ephemeral disks.
The primary change here is updating ex_create_multiple_nodes to create disks and
nodes in parallel.

wrigri committed:
Add functionality to create_node & create_volume to use existing
disks of the same name.
Also some pep8 fixes.

Commit:
[LIBCLOUD-434] Implement iterate_* methods in the Route53 driver.

This way all the records are returned and not just the ones which fit onto the
first page.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update CHANGES.

virtuald committed:
[LIBCLOUD-455] Don't throw if 'imageId' attribute is not present in the
list_nodes response.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Re-order methods on the base NodeDriver object so they are consistent with other
APIs ("public" method first, "private" method at the end, etc.).

Kami committed:
Add new KeyPair class and methods for the key pair management to the NodeDriver
class.

Kami committed:
Update existing compute drivers (OpenStack, CloudStack, EC2) which expose key
pair management through the extension methods to expose it through a new
standard key pair management API.

Also update affected code and tests, deprecate existing extension methods and
add some missing tests.

Note: Old and now deprecated methods will work until a next major release.

Kami committed:
Update docs.

Kami committed:
Correctly handling missing 'image' attribute in the OpenStack list nodes
response and add a test case for it.

Part of LIBCLOUD-455.

Kami committed:
Update CHANGES.

Kami committed:
docs: Update upgrade notes,  add a new section about key pair management
and update existing docs.

Kami committed:
Add and implement get_key_pair method.

Also add tests and update documentation.

Kami committed:
Re-generate documentation fixtures.

Kami committed:
Modify fixture generation script to treat list_records and list_zones as
implemented if iterate_records and iterate_zones method is implemented.

Also fix the script to preserver correct order.

Kami committed:
Don't include local storage provider in the supported providers table.

Kami committed:
Re-generate supported providers and supported provider methods table.

Kami committed:
Update .gitignore.

Kami committed:
[LIBCLOUD-457] Reset context on the Connection class after the request has
finished.

Also update affected Abiquo driver to use "context" attribute correctly and
not rely on it being present across multiple requests. Instead of the
"context", use a new attribute called "cache".

Kami committed:
Update generate fixture script to treat list_* method as implemented if the
corresponding iterate_* method is implemented.

Kami committed:
Update Nimbus.io storage driver name so it's unique.

Kami committed:
Re-generate documentation fixtures.

Kami committed:
Give better "name" to ELB driver, re-generate documentation fixtures.

Kami committed:
Add some missing docstring annotations.

Kami committed:
Remove NODE_STATE_MAP attribute from Response class. No idea how this got here.

Kami committed:
Add some more comments and missing docstrings.

Kami committed:
docs: Update troubleshooting section.

Kami committed:
Update fixture generation script to include key pair management methods in the
resulting fixtures.

Kami committed:
docs: Include key pair management supported methods table.

Kami committed:
docs: re-generate fixtures.

Kami committed:
Add more information to the readme and rename a file to README.rst.

Kami committed:
Migrate CONTRIBUTING.md file to reST format.

Kami committed:
Update MANIFEST.in.

Kami committed:
docs: Update rackspace driver docs.

Kami committed:
docs: Fix typo in upgrade notes.

Kami committed:
Refactor libcloud.common.base.Connection to avoid doing duplicate work and
remove some additional code.

Also update affected code.

Kami committed:
docs: Update "SSL Certificate Validation" section.

Kami committed:
Update a link to point to the correct page.

Kami committed:
Remove support for old and deprecated VERIFY_SSL_STRICT variable.

This variable was only used in an old version when we enabled cert validation
to ease the migration path. It has been deprecated since then.

Kami committed:
Set LibcloudHTTPSConnection.verify variable to True by default.

Kami committed:
Use unittest2 and assertRaisesRegexp instead.

Kami committed:
Remove unused import.

Kami committed:
Update upgrade notes.

Kami committed:
docs: Also document VolumeSnapshot class.

Kami committed:
docs: Use ":class:" when referencing a class in the terminology section.

Kami committed:
Update __all__ in libcloud.compute.base module.

Kami committed:
docs: Update "SSL Certificate Validation".

Kami committed:
docs: Re-order sections.

Kami committed:
Make 'DEFAULT_CONTENT_TYPE' a constant and add a test case for it.

Kami committed:
Update CHANGES.

Kami committed:
Add "strict_mode" to the storage driver.

When this mode is enabled (disabled by default), it will default to an old
behavior when no content type is provided and none can't be automatically
detected.

Kami committed:
docs: Update upgrade notes.

Kami committed:
Use unittest2.

Kami committed:
Merge branch 'promote_key_pair_management_to_base_compute_api' into trunk

Kami committed:
docs: re-organize upgrade notes, move non breaking stuff towards the end.

Kami committed:
Add badges to the readme.

Kami committed:
docs: fix underline.

Kami committed:
Update CloudStack driver to better handle errors and throw ProviderError
instead of a generic Exception.

Commit:
Issue LIBCLOUD-459: Support region without zone in GCE.

Patch by Charles Walker <charles dot walker dot 37 at gmail dot com>.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Fix a typo.

sebgoa committed:
docs: Add documentation for CloudStack driver.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Add bash script for verifying / comparing md5 checksums of the release
artifacts on the Apache and PyPi servers.

Kami committed:
Fix a typo in the docs.

Kami committed:
docs: Update "releasing" section.

Commit:
Minor spelling fix.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Modify ex_list_networks methods in CloudStack driver to not thrown if there
are no networks available.

Kami committed:
docs: Promote some CloudStack code snippets to full blown code examples which
can be run directly and add missing imports and variables.

Kami committed:
docs: Fix a typo, it should be "networks", not "network".

Kami committed:
docs: Fix some typos and formatting.

Kami committed:
Re-generate documentation fixtures.

Kami committed:
docs: Update example, no need to disable cert validation.

sebgoa committed:
LIBCLOUD-439: Be consistent with other drivers and use underscores in the extra
dictionary attribute names.

Kami committed:
docs: Stub out Exoscale and Ikoula driver docs.

alex committed:
SLicehost isn't a thing anymore

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

wrigri committed:
Test updates for v1 plus additional changes:
  Add snapshot support
  Add .get() to many locations to make the conversion from json dictionary to
  libcloud objects more robust.

Kami committed:
Make script executable.

Kami committed:
Re-generate documentation fixtures.

Kami committed:
docs: Add costrings conventions section (moved from the site).

gigimon committed:
For convenience, add ex_create_port_forwarding_rule and
ex_delete_port_forwarding_rule method to the CloudStackNode class.

For consistency, also rename "ex_add_ip_forwarding_rule" to
"ex_create_ip_forwarding_rule".

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

wrigri committed:
Add Snapshots to demo code and a snapshot fix

wrigri committed:
Fixes for dealing with disk on multiple node creation/destruction.

wrigri committed:
Tests for Snapshots (plus fix for list_targetpools test)

wrigri committed:
Lint fixes

wrigri committed:
One more lint fix

Kami committed:
Bump API version used by EC2 driver to 2013-10-15 and update affected test
fixtures.

Kami committed:
Add ex_get_limits method to EC2 driver.

Kami committed:
Update CHANGES.

wrigri committed:
Initialize PyCrypto's RNG
(We don't actually use the RNG, but this should prevent certain errors if
libcloud is running in a fork of an application that has already imported the
Crypto library.)

Kami committed:
Update us-west-1 region in the EC2 driver to include c3 instance types.
Also include pricing information.

Reference: https://twitter.com/awscloud/status/412873932828442624

Kami committed:
Update CHANGES.

Kami committed:
Update CHANGES file to use a more consistent style. This way file can more
easily be processed automatically by different tools.

Kami committed:
Update CHANGES.

Kami committed:
Add a script for generating a Markdown formatted list of contributors from the
CHANGES file.

wrigri committed:
Add support for reading/setting node scheduling options.

Kami commented on libcloud/common/google.py:
```diff
@@ -82,6 +82,8 @@
     from Crypto.PublicKey import RSA
     from Crypto.Signature import PKCS1_v1_5
+    import Crypto.Random
+    Crypto.Random.atfork()
```
Hm, we aren't forking anywhere, but if this reseeds the random generator and adds more entropy to the pool, explicitly calling it here is not such a bad idea.

Also, no wonder people don't like pycrypto. This seems like a pretty awful API for reseeding the RNDG.

wrigri replied:
I think the issue is that even though libcloud isn't forking anywhere, if someone else is using pycrypto and then forks to call libcloud, it can cause an error.  This is basically a safety call that just makes sure that libcloud is working from a fresh state.  (There may be a better way to do it, but this seemed to be the general suggestion when I was searching for the particular error seen by a coworker working on some different third-party code.)

Kami replied:
OK, thanks for the clarification. That's kinda what I thought.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -941,6 +962,20 @@ def list_sizes(self, location=None):
                 list_sizes = [self._to_node_size(s) for s in response['items']]
         return list_sizes
 
+    def ex_list_snapshots(self):
```
Hm, this lists the snapshots for the persistent disks, right?

Compute part of the API does include some standard methods and classes for managing block storage volumes and snapshots. `list_volume_snapshots` is one of the supported methods.

The problem is that those methods refer to the block storage volumes which are not the same as instance storage (or persistent disks in this case), so I'm not yet sure what we should do here.

I need to have a better look and think some more about it. I just want to make sure this whole thing doesn't confuse the user.

wrigri replied:
I'm open for suggestions on this.  I saw the list_volume_snapshots as well, but those were attributes of the storage volumes.  For GCE, snapshots are global resources that may just happen to have been created from a specific disk.

Kami replied:
OK, in this case we should leave this extension method as it is.

One thing I'm wondering is - can you filter snapshots for a specific volume? If you can, it would make sense to also implement `list_volume_snapshots` from the base API.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1447,27 +1510,76 @@ def ex_create_multiple_nodes(self, base_name, size, image, number,
-                        node_list[i] = GCEFailedNode(name, error, code)
+            time.sleep(2)
+            for i, status in enumerate(status_list):
+                # If disk does not yet exist, check on its status
```
Is it possible to refactor some of the functionality outside of this loop into multiple smaller methods?

Currently it's really dense and hard to read / follow.

wrigri replied:
Yeah, I'm not 100% happy with it either, so I'll give it another pass and see if I can improve it.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1447,27 +1510,76 @@ def ex_create_multiple_nodes(self, base_name, size, image, number,
+                    try:
+                        response = self.connection.request(
+                            status['disk_response']['selfLink']).object
+                    except:
```
What exception is thrown here? Can we catch a more specific one?

Even if we cant, you should use `except Exception`.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1447,27 +1510,76 @@ def ex_create_multiple_nodes(self, base_name, size, image, number,
-                    name = '%s-%03d' % (base_name, i)
-                    if error:
-                        node_list[i] = GCEFailedNode(name, error, code)
+            time.sleep(2)
```
Is it possible to make this delay configurable?

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1557,18 +1672,82 @@ def create_volume(self, size, name, location=None, image=None,
+
+        return request, volume_data, params
+
+    def create_volume(self, size, name, location=None, image=None,
```
Standard API looks like this:

``` python
def create_volume(self, size, name, location=None, snapshot=None):
```

If a user would use regular and not keyword arguments, this method would not work as expected.

To make it compatible with the base API, please change the signature to:

``` python
def create_volume(self, size, name, location=None, snapshot=None, image=None, use_existing=True):
```

wrigri commented:
@Kami Thanks for the review.  I replied to a few specific comments, but overall, I don't disagree with any of your feedback, so I'll work on some improvements.

Kami committed:
Call title on the contributor name so all the names follow the same format.

Kami committed:
Update CHANGES.

Kami committed:
Modify lint target to also lint files in contrib/.

Kami commented:
@wrigri Sounds good. Thanks.

@wrigri @erjohnso It would also be good to add some GCE specific docs with examples to Libcloud (https://github.com/apache/libcloud/tree/trunk/docs). Then you can also add a link to GCE salt cloud provisioner there :)

Kami committed:
To make it more clear, use keyword arguments.

Kami committed:
Update CHANGES.

Kami committed:
Use consistent docstring format.

Kami committed:
Use sphinxcontrib.fulltoc sphinx extension.

Kami committed:
Fix dependency name.

Kami committed:
docs: Add a missing section, fix some formatting issues.

Kami committed:
docs: Fix a bunch of build warnings.

Kami committed:
Add zip_safe flag to setup.py.

Kami committed:
Update MANIFEST.in.

Kami committed:
Use my fork of fulltoc sphinx extension.

Kami committed:
docs: Try a nasty RTD hack.

Kami committed:
Revert the fulltoc madness.

Kami committed:
docs: Fix a typo.

Kami committed:
docs: Move changelog and upgrade notes to bottom of the page.

Kami committed:
Allow filterting contributors by release version.

Commit:
Issue LIBCLOUD-464: Add methods for deleting security groups to the EC2 driver.

Kami committed:
Add --include-tickets option to generate contributors list script.

Commit:
Issue LIBCLOUD-465: Add new EC2 I2 instance types to EC2 driver.

Also update pricing information.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Commit:
Issue: LIBCLOUD-463 Add VPC support for security group creation (EC2 driver)

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update CHANGES.

Kami committed:
Update CHANGES ordering.

Kami committed:
Update CHANGES.

Commit:
Issue: LIBCLOUD-466 Add ingress/egress VPC support to security group rules.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Commit:
Issue: LIBCLOUD-477 Add list/delete/create network (VPC) calls within EC2

wrigri committed:
Fix signature for create_volume to align with the base API

wrigri committed:
Catch some additional errors from the authentication.
Also, update the docs to reflect the new console.

wrigri committed:
flake8 fixes

Kami committed:
docs: Fix a typo.

Kami committed:
Move "private" method to the end, add a docstrings.

Kami committed:
Update CHANGES.

Kami committed:
Fix indentation.

Kami committed:
Fix test failure in Python 2.5; enumerate in 2.5 doesn't support start argument.

Kami committed:
Fix ex_delete_security_group method to call the correct method.

Kami committed:
Fix a typo in the test and make sure the tests actually run.

Kami committed:
Update CHANGES.

cderamus committed:
Merge pull request #2 from apache/trunk

Pulling in latest updates

Commit:
Issue LIBCLOUD-469: Add Reserved Instance support to EC2 driver.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Commit:
Changed the return type to a list of EC2Network objects
Updated list/delete network tests

Kami committed:
Modify EC2ResercedNode to call parent constructor and drop unused arguments from
the constructor.

Kami committed:
Also pass "size" attribute to EC2ReservedNode constructor.

Kami committed:
Move test from NimbusTests to EC2Tests class.

Kami committed:
list_reserved_nodes -> ex_list_reserved_nodes, update docstrings.

Kami committed:
Update CHANGES.

cderamus committed:
Merge pull request #3 from apache/trunk

Pulling in latest changes

Commit:
Merge branch 'LIBCLOUD-467_Add_VPC_Lifecycle_Support' into trunk

Conflicts:
	libcloud/compute/drivers/ec2.py
	libcloud/test/compute/test_ec2.py

Commit:
Updated ex_create_network to return an EC2Network object
Updated test_ex_create_network accordingly

Commit:
Destroy -> Delete

Kami committed:
Modify ex_delete_network method to take "vpc" as an argument.

Kami committed:
Update CHANGES.

Kami committed:
Add back a comment which was while resolving merge conflicts.

Commit:
LIBCLOUD-468: Add VPC subnet calls to the EC2 driver.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Modify ex_delete_subnet method to take a EC2NetworkSubnet instance instead.

Kami committed:
Update CHANGES.

Commit:
Issue LIBCLOUD-471: Add ex_get_console_output method to the EC2 driver.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update docstring - remove obsolete comment, clarify which keys the dictionary
contains.

Kami committed:
Fix test failure under Python 3.

Commit:
Issue LIBCLOUD-473: Add more EBS attributes to the "extra" dictionary in the
StorageVolume class.

Also refactor code to use new "_get_resource_tags" method.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update CHANGES.

Kami committed:
Update docstrings, fix lint issue which arose during the conflict resolution.

Kami committed:
Update CHANGES.

Kami committed:
docs: Update contribution guide, add note about git format-patch and commit
squashing.

Kami committed:
docs: Update contribution page - add "code conventions" section.

Kami committed:
docs: Add a note about synchronizing the branch.

gigimon committed:
LIBCLOUD-475 fix key_name in Ec2 and Cloudstack drivers

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update CHANGES.

Kami committed:
Fix lint issue.

Commit:
Issue LIBCLOUD-477: Add extra utility function into ec2 driver

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update docstring.

Kami committed:
cast_func -> transform_func.

Commit:
Issue LIBCLOUD-474: Add Network Interface Support to the EC2 driver.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Document some choices in the ssh client.

Kami committed:
Fix a deployment buffering issue spotted by @pquerna which could happen if the
the executed command produced a lot of output.

Kami committed:
Update CHANGES.

Kami committed:
Add short delay to prevent busy waiting.

Commit:
Issue LIBCLOUD-480: Add additional attributes to the "extra" dictionary of
the VolumeSnapshot object in the EC2 driver.

Also modify create_volume_snapshot method to correctly handle "name" argument.

Previous, "name" argument was used as a snapshot description, now it's used
as a Tag with a key "Name".

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Cleanup and improve some docstrings.

Kami committed:
Modify _connect_and_run_deployment_script to return a Node object on success.

Kami committed:
Use better variable names, move check to the top.

Commit:
Issue LIBCLOUD-481: Store additional attributes (iops, tags,
block_device_mapping) in the   "extra" dictionary of the NodeImage object in
the EC2 driver.

Also fix ex_image_ids filtering in the list_images method.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Update Contributing document.

Kami committed:
Update CHANGES.

Kami committed:
Update CHANGES to include github pull request numbers.

wrigri committed:
Split up ex_create_multiple_nodes to make it a bit easier to read.
Also, add poll_interval keyword.

wrigri commented:
@Kami I think I addressed all of your feedback.  Please take a look and let me know if you have other feedback or suggestions.  Also, let me know if you have any thoughts on the list_snapshots thing.

Kami committed:
Update CHANGES.

Kami committed:
docs: Add information on how to instantiate a HostVirtual driver.

Kami committed:
Re-generate documentation fixtures.

Kami committed:
Add allow_insecure=True attribute to the HostVirtual connection class.

Kami committed:
Update docstrings.

Kami committed:
Update HostVirtual compute driver to use super and take in all the supported
arguments.

Kami committed:
Update HostVirtual DNS driver constructor to only take in "key" and other
supported arguments. Previously it also took "secret" which it silently
ignored.

Part of LIBCLOUD-483.

Kami committed:
Update CHANGES.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1982,22 +2299,31 @@ def ex_destroy_network(self, network):
+            node.extra['boot_disk'].destroy()
         return True
 
     def ex_destroy_multiple_nodes(self, nodelist, ignore_errors=True,
```
For consistency please use `node_list` instead of `nodelist`.

Kami commented:
The patch is pretty big, but I think it mostly looks good.

Only thing I would ask you is if you could re-organize the methods a bit - this means moving "private" (methods prefixed with an underscore) after the "public" method to the bottom (https://libcloud.readthedocs.org/en/latest/development.html#function-and-method-ordering).

This will make the code and Sphinx documentation generated from the docstrings a bit easier to read and follow.

Kami committed:
docs: Add a paragraph about **kwargs to the code coventions section.

wrigri committed:
Rename nodelist to node_list

Commit:
Issue LIBCLOUD-485: Added a new constant dictionary,
RESOURCE_EXTRA_ATTRIBUTES_MAP, at the top of the driver which will be used to
store the xpath and transform defintions for each element of the extra
dictionary for each resource. This functionality previously existed in the
private _to_XXXX methods, but was re-located for readibility and to make it
centralize the functionality for easier management and updates going forward.

Signed-off-by: Tomaz Muraus <tomaz@apache.org>

Kami committed:
Send 'scrub_data' parameter when destroying a DigitalOcean node.

Part of LIBCLOUD-487.

Kami committed:
Send a query parameter wtih a value of '1'. API doesn't seem to consider string
'true' as true.

wrigri committed:
Reorder methods to put them in public/private/internal order.

wrigri committed:
Add list_volume_snapshots method to help fill out the API.
This is not an efficient method, but in GCE snapshots are not keyed on the volumes they are created from.

wrigri commented:
@Kami I reordered the methods in gce.py, and made that minor change to nodelist.

Also, I went ahead and added list_volume_snapshots.  It turns out it wasn't too hard to filter by the volume they were created from.  If the volume is deleted, the volume reference is removed from the snapshot, but that shouldn't affect list_volume_snapshots negatively.

Kami committed:
Modify verify checksum script to also check checksum of zip artifact.

Kami committed:
Modify script to return error if the file doesn't exist.

Kami committed:
To be on the safe side, add some asserts to the script.

Kami committed:
docs: fix a typo.

Kami committed:
docs: update formatting.

Kami committed:
Add some more test for verify_hostname.

Kami committed:
Updates CHANGES frm 0.13.3 branch.

Kami committed:
docs: Update Committer Guide chapter.

Kami commented:
@wrigri Looks good, thanks.

Can you please sync your branch with the latest trunk and squash the commits? After you do this, I will go ahead and merge the patch.

Kami committed:
Update CHANGES.

wrigri committed:
Merge branch 'LIBCLOUD-450_GCEv1' of https://github.com/wrigri/libcloud into LIBCLOUD-450_GCEv1

Kami commented:
Hm, the pull request looks really messed up now.

I've tried manually checking out your branch and squashing the commits, but I also encountered conflicts. To save both of us some time & pain, I will make an exception this time and directly merge your branch into trunk.

Kami commented:
I've merged your branch into trunk - thanks!

Also, as we've talked in the past, it would now be great to add some GCE specific documentation to Libcloud docs.

Kami commented:
@wrigri Ugh, I just wanted to update the JIRA issue and I noticed you have uploaded a squashed patch there.

Sorry for that, I have totally missed this before.
